File: testdata/fixtures/src/sample.ex
# sample.ex — comprehensive Elixir syntax fixture for parser testing.
# Covers: modules, structs, protocols, behaviours, macros, pattern matching,
# pipe operator, recursion, comprehensions, processes, GenServer, tasks,
# typespecs, guards, with expressions, streams, ETS.

defmodule Sample do
  @moduledoc """
  Comprehensive Elixir syntax fixture covering idiomatic patterns.
  """

  # -------------------------------------------------------------------------- #
  # Module attributes (constants)
  # -------------------------------------------------------------------------- #

  @max_retries 3
  @default_timeout_ms 30_000
  @version "1.0.0"

  # -------------------------------------------------------------------------- #
  # Typespecs
  # -------------------------------------------------------------------------- #

  @type uuid :: String.t()
  @type status :: :pending | :running | :done | :failed
  @type result(t) :: {:ok, t} | {:error, String.t()}

  # -------------------------------------------------------------------------- #
  # Structs
  # -------------------------------------------------------------------------- #

  defmodule User do
    @moduledoc "Domain entity representing an application user."

    @enforce_keys [:id, :name, :email]
    defstruct [:id, :name, :email, role: "member", status: :pending]

    @type t :: %__MODULE__{
            id:     String.t(),
            name:   String.t(),
            email:  String.t(),
            role:   String.t(),
            status: Sample.status()
          }

    @spec new(String.t(), String.t(), String.t()) :: {:ok, t()} | {:error, String.t()}
    def new(id, name, email) do
      cond do
        String.trim(name) == "" -> {:error, "name must not be blank"}
        not String.contains?(email, "@") -> {:error, "email '#{email}' is invalid"}
        true -> {:ok, %__MODULE__{id: id, name: name, email: email}}
      end
    end
  end

  defmodule Point do
    @moduledoc "2-D geometric point."

    defstruct [:x, :y]

    @type t :: %__MODULE__{x: float(), y: float()}

    def origin, do: %__MODULE__{x: 0.0, y: 0.0}

    def distance(%__MODULE__{x: x1, y: y1}, %__MODULE__{x: x2, y: y2}) do
      :math.sqrt(:math.pow(x1 - x2, 2) + :math.pow(y1 - y2, 2))
    end

    defimpl String.Chars do
      def to_string(%{x: x, y: y}), do: "(#{x}, #{y})"
    end
  end

  # -------------------------------------------------------------------------- #
  # Protocols
  # -------------------------------------------------------------------------- #

  defprotocol Describable do
    @doc "Returns a human-readable description of the value."
    @spec describe(t()) :: String.t()
    def describe(value)
  end

  defimpl Describable, for: User do
    def describe(%User{name: name, email: email}),
      do: "User(#{name}, #{email})"
  end

  defimpl Describable, for: Point do
    def describe(%Point{x: x, y: y}), do: "Point(#{x}, #{y})"
  end

  # -------------------------------------------------------------------------- #
  # Behaviours
  # -------------------------------------------------------------------------- #

  defmodule Repository do
    @moduledoc "Repository port for persistence."

    @callback find_by_id(id :: String.t()) :: {:ok, term()} | {:error, :not_found}
    @callback list() :: list(term())
    @callback save(entity :: term()) :: {:ok, term()} | {:error, String.t()}
    @callback delete(id :: String.t()) :: :ok | {:error, :not_found}
  end

  # -------------------------------------------------------------------------- #
  # Pattern matching & guards
  # -------------------------------------------------------------------------- #

  @spec classify(term()) :: String.t()
  def classify(nil),                           do: "null"
  def classify(b) when is_boolean(b),          do: "boolean: #{b}"
  def classify(n) when is_integer(n) and n < 0, do: "negative int #{n}"
  def classify(n) when is_integer(n),           do: "non-negative int #{n}"
  def classify(s) when is_binary(s) and s == "", do: "empty string"
  def classify(s) when is_binary(s),            do: "string of length #{String.length(s)}"
  def classify([]),                             do: "empty list"
  def classify(l) when is_list(l),              do: "list of #{length(l)}"
  def classify(%{__struct__: mod}),             do: "struct #{mod}"
  def classify(_),                              do: "unknown"

  # -------------------------------------------------------------------------- #
  # Recursion & tail-call optimisation
  # -------------------------------------------------------------------------- #

  @spec factorial(non_neg_integer()) :: non_neg_integer()
  def factorial(n) when n >= 0, do: factorial(n, 1)

  defp factorial(0, acc), do: acc
  defp factorial(n, acc), do: factorial(n - 1, n * acc)

  @spec fibonacci(non_neg_integer()) :: [non_neg_integer()]
  def fibonacci(n), do: fib_acc(n, 0, 1, [])

  defp fib_acc(0, _, _, acc), do: Enum.reverse(acc)
  defp fib_acc(n, a, b, acc), do: fib_acc(n - 1, b, a + b, [a | acc])

  # -------------------------------------------------------------------------- #
  # Pipe operator & Enum
  # -------------------------------------------------------------------------- #

  @spec word_frequency(String.t()) :: %{String.t() => non_neg_integer()}
  def word_frequency(text) do
    text
    |> String.downcase()
    |> String.split(~r/\W+/, trim: true)
    |> Enum.frequencies()
  end

  @spec top_n(map(), non_neg_integer()) :: [{String.t(), non_neg_integer()}]
  def top_n(freq, n) do
    freq
    |> Enum.sort_by(fn {_k, v} -> v end, :desc)
    |> Enum.take(n)
  end

  # -------------------------------------------------------------------------- #
  # For comprehensions
  # -------------------------------------------------------------------------- #

  @spec pythagorean(pos_integer()) :: [{pos_integer(), pos_integer(), pos_integer()}]
  def pythagorean(limit) do
    for c <- 1..limit,
        b <- 1..c,
        a <- 1..b,
        a * a + b * b == c * c,
        do: {a, b, c}
  end

  # -------------------------------------------------------------------------- #
  # with expression
  # -------------------------------------------------------------------------- #

  @spec process_user(map()) :: result(User.t())
  def process_user(params) do
    with {:ok, id}    <- Map.fetch(params, "id"),
         {:ok, name}  <- Map.fetch(params, "name"),
         {:ok, email} <- Map.fetch(params, "email"),
         {:ok, user}  <- User.new(id, name, email) do
      {:ok, user}
    else
      :error          -> {:error, "missing required field"}
      {:error, msg}   -> {:error, msg}
    end
  end

  # -------------------------------------------------------------------------- #
  # Streams (lazy evaluation)
  # -------------------------------------------------------------------------- #

  @spec fibonacci_stream() :: Enumerable.t()
  def fibonacci_stream do
    Stream.unfold({0, 1}, fn {a, b} -> {a, {b, a + b}} end)
  end

  @spec process_large_file(String.t()) :: Enumerable.t()
  def process_large_file(path) do
    path
    |> File.stream!()
    |> Stream.map(&String.trim/1)
    |> Stream.reject(&(&1 == ""))
    |> Stream.with_index(1)
  end

  # -------------------------------------------------------------------------- #
  # Macros
  # -------------------------------------------------------------------------- #

  defmacro assert_ok(expr) do
    quote do
      case unquote(expr) do
        {:ok, value} -> value
        {:error, reason} -> raise "Expected {:ok, _}, got {:error, #{inspect(reason)}}"
      end
    end
  end

  # -------------------------------------------------------------------------- #
  # GenServer
  # -------------------------------------------------------------------------- #

  defmodule Counter do
    @moduledoc "Simple stateful counter implemented as a GenServer."

    use GenServer

    @spec start_link(non_neg_integer()) :: GenServer.on_start()
    def start_link(initial \\ 0) do
      GenServer.start_link(__MODULE__, initial, name: __MODULE__)
    end

    @spec increment(pid() | atom()) :: :ok
    def increment(pid \\ __MODULE__), do: GenServer.cast(pid, :increment)

    @spec value(pid() | atom()) :: non_neg_integer()
    def value(pid \\ __MODULE__),     do: GenServer.call(pid, :value)

    @spec reset(pid() | atom()) :: :ok
    def reset(pid \\ __MODULE__),     do: GenServer.cast(pid, :reset)

    # --- callbacks ---

    @impl true
    def init(initial), do: {:ok, initial}

    @impl true
    def handle_cast(:increment, count), do: {:noreply, count + 1}
    def handle_cast(:reset, _),         do: {:noreply, 0}

    @impl true
    def handle_call(:value, _from, count), do: {:reply, count, count}
  end

  # -------------------------------------------------------------------------- #
  # Tasks (async)
  # -------------------------------------------------------------------------- #

  @spec fetch_all(list(String.t())) :: list(term())
  def fetch_all(urls) do
    urls
    |> Enum.map(&Task.async(fn -> fetch_one(&1) end))
    |> Task.await_many(@default_timeout_ms)
  end

  defp fetch_one(url) do
    {:ok, "data from #{url}"}
  end

  # -------------------------------------------------------------------------- #
  # ETS (Erlang Term Storage)
  # -------------------------------------------------------------------------- #

  @spec create_cache(atom()) :: :ets.tid()
  def create_cache(name) do
    :ets.new(name, [:set, :public, :named_table, read_concurrency: true])
  end

  @spec cache_put(:ets.tid(), term(), term()) :: true
  def cache_put(table, key, value) do
    :ets.insert(table, {key, value, System.monotonic_time()})
  end

  @spec cache_get(:ets.tid(), term()) :: {:ok, term()} | :miss
  def cache_get(table, key) do
    case :ets.lookup(table, key) do
      [{^key, value, _ts}] -> {:ok, value}
      [] -> :miss
    end
  end

  # -------------------------------------------------------------------------- #
  # Exception handling
  # -------------------------------------------------------------------------- #

  @spec safe_divide(number(), number()) :: result(float())
  def safe_divide(_, 0), do: {:error, "division by zero"}
  def safe_divide(a, b), do: {:ok, a / b}

  def with_rescue(fun) do
    try do
      {:ok, fun.()}
    rescue
      e in [ArithmeticError, ArgumentError] -> {:error, Exception.message(e)}
    catch
      :throw, reason -> {:error, "thrown: #{inspect(reason)}"}
    end
  end
end

# -------------------------------------------------------------------------- #
# Entry point (when run as script)
# -------------------------------------------------------------------------- #

case Sample.User.new("u1", "Alice", "alice@example.com") do
  {:ok, user}    -> IO.inspect(user, label: "user")
  {:error, msg}  -> IO.puts("Error: #{msg}")
end

IO.inspect(Sample.fibonacci(10), label: "fibonacci")
IO.inspect(Sample.pythagorean(20), label: "pythagorean")

p1 = %Sample.Point{x: 3.0, y: 4.0}
p2 = Sample.Point.origin()
IO.puts("Distance: #{Sample.Point.distance(p1, p2)}")

File: testdata/fixtures/src/sample.scala
/**
 * sample.scala — comprehensive Scala 3 syntax fixture for parser testing.
 * Covers: case classes, enums, traits, given/using, extension methods,
 * type classes, opaque types, union/intersection types, match types,
 * for-comprehensions, futures, pattern matching, higher-kinded types.
 */

import scala.concurrent.{ExecutionContext, Future}
import scala.util.{Failure, Success, Try}

// -------------------------------------------------------------------------- //
// Opaque types
// -------------------------------------------------------------------------- //

opaque type Uuid = String
object Uuid:
  def apply(s: String): Uuid = s
  extension (id: Uuid) def value: String = id

opaque type PositiveInt <: Int = Int
object PositiveInt:
  def apply(n: Int): Either[String, PositiveInt] =
    if n > 0 then Right(n) else Left(s"Expected positive, got $n")

// -------------------------------------------------------------------------- //
// Enums (Scala 3)
// -------------------------------------------------------------------------- //

enum Status:
  case Pending, Running, Done, Failed

  def isTerminal: Boolean = this == Done || this == Failed
  def label: String = this match
    case Pending => "Waiting"
    case Running => "In progress"
    case Done    => "Completed"
    case Failed  => "Failed"

enum Result[+T]:
  case Success(value: T)
  case Failure(error: Throwable)
  case Loading

  def map[U](f: T => U): Result[U] = this match
    case Success(v) => Success(f(v))
    case Failure(e) => Failure(e)
    case Loading    => Loading

  def getOrElse[U >: T](default: => U): U = this match
    case Success(v) => v
    case _          => default

// -------------------------------------------------------------------------- //
// Case classes & companion objects
// -------------------------------------------------------------------------- //

case class Point(x: Double, y: Double):
  def distanceTo(other: Point): Double =
    math.sqrt(math.pow(x - other.x, 2) + math.pow(y - other.y, 2))

  def +(other: Point): Point = Point(x + other.x, y + other.y)
  def *(scale: Double): Point = Point(x * scale, y * scale)

object Point:
  val origin: Point = Point(0, 0)

case class User(
  id: Uuid,
  name: String,
  email: String,
  role: String = "member"
)

// -------------------------------------------------------------------------- //
// Traits & type classes
// -------------------------------------------------------------------------- //

trait Repository[F[_], T]:
  def findById(id: Uuid): F[Option[T]]
  def list: F[List[T]]
  def save(entity: T): F[T]
  def delete(id: Uuid): F[Boolean]

trait Show[T]:
  def show(value: T): String

trait Eq[T]:
  def eqv(a: T, b: T): Boolean

trait Ord[T] extends Eq[T]:
  def compare(a: T, b: T): Int
  def eqv(a: T, b: T): Boolean = compare(a, b) == 0
  def lt(a: T, b: T): Boolean  = compare(a, b) < 0
  def gt(a: T, b: T): Boolean  = compare(a, b) > 0

// -------------------------------------------------------------------------- //
// Given instances (type class instances)
// -------------------------------------------------------------------------- //

given Show[Int] with
  def show(n: Int): String = n.toString

given Show[String] with
  def show(s: String): String = s""""$s""""

given Show[User] with
  def show(u: User): String = s"User(${u.name}, ${u.email})"

given [T: Show]: Show[List[T]] with
  def show(xs: List[T]): String =
    xs.map(summon[Show[T]].show).mkString("[", ", ", "]")

given Ord[Int] with
  def compare(a: Int, b: Int): Int = java.lang.Integer.compare(a, b)

given Ord[String] with
  def compare(a: String, b: String): Int = a.compareTo(b)

// -------------------------------------------------------------------------- //
// Extension methods
// -------------------------------------------------------------------------- //

extension [T: Show](value: T)
  def display: String = summon[Show[T]].show(value)

extension (s: String)
  def isValidEmail: Boolean = s.contains('@') && s.contains('.')
  def camelToSnake: String =
    "[A-Z]".r.replaceAllIn(s, m => s"_${m.matched.toLowerCase}").stripPrefix("_")

extension [T](xs: List[T])
  def safeHead: Option[T] = xs.headOption
  def safeLast: Option[T] = xs.lastOption
  def groupByKey[K](f: T => K): Map[K, List[T]] = xs.groupBy(f)

// -------------------------------------------------------------------------- //
// Context functions (using parameters)
// -------------------------------------------------------------------------- //

def showAll[T: Show](items: List[T]): String =
  items.map(_.display).mkString(", ")

def sortWith[T: Ord](xs: List[T]): List[T] =
  val ord = summon[Ord[T]]
  xs.sortWith((a, b) => ord.lt(a, b))

// -------------------------------------------------------------------------- //
// Union & intersection types
// -------------------------------------------------------------------------- //

type StringOrInt = String | Int
type Named = { def name: String }

def stringify(v: StringOrInt): String = v match
  case s: String => s
  case n: Int    => n.toString

// -------------------------------------------------------------------------- //
// Match types
// -------------------------------------------------------------------------- //

type Elem[T] = T match
  case List[t]  => t
  case Option[t] => t
  case _         => T

// -------------------------------------------------------------------------- //
// For-comprehensions & Option/Either
// -------------------------------------------------------------------------- //

def safeDivide(a: Int, b: Int): Either[String, Int] =
  if b == 0 then Left(s"Division by zero: $a / $b")
  else Right(a / b)

def computeRatio(x: Int, y: Int, z: Int): Either[String, Double] =
  for
    sum  <- safeDivide(x + y, 2)
    prod <- safeDivide(sum, z)
  yield prod.toDouble

// -------------------------------------------------------------------------- //
// Pattern matching (advanced)
// -------------------------------------------------------------------------- //

def describe(value: Any): String = value match
  case null              => "null"
  case n: Int if n < 0  => s"negative int $n"
  case n: Int           => s"non-negative int $n"
  case s: String        => s"string(${s.length})"
  case Point(0, 0)      => "origin"
  case Point(x, y)      => s"point($x, $y)"
  case xs: List[?]      => s"list of ${xs.size} items"
  case Some(v)          => s"some($v)"
  case None             => "none"
  case _                => s"unknown: ${value.getClass.getSimpleName}"

// -------------------------------------------------------------------------- //
// Futures
// -------------------------------------------------------------------------- //

def fetchData(url: String)(using ec: ExecutionContext): Future[String] =
  Future { s"data from $url" }

def fetchAll(urls: List[String])(using ec: ExecutionContext): Future[List[String]] =
  Future.sequence(urls.map(fetchData))

// -------------------------------------------------------------------------- //
// Lazy val & higher-kinded
// -------------------------------------------------------------------------- //

lazy val expensiveValue: Int = {
  println("computed!")
  42
}

def lift[F[_], A, B](f: A => B)(using ev: F[A] => F[B]): F[A] => F[B] = ev

// -------------------------------------------------------------------------- //
// Fibonacci via infinite LazyList
// -------------------------------------------------------------------------- //

lazy val fibonacci: LazyList[Long] =
  def go(a: Long, b: Long): LazyList[Long] = a #:: go(b, a + b)
  go(0, 1)

// -------------------------------------------------------------------------- //
// Entry point
// -------------------------------------------------------------------------- //

@main def run(): Unit =
  val user = User(Uuid("u1"), "Alice", "alice@example.com")
  println(user.display)

  val fibs = fibonacci.take(10).toList
  println(fibs.display)

  val p = Point(3, 4)
  println(s"Distance: ${p.distanceTo(Point.origin)}")

  println(sortWith(List(3, 1, 4, 1, 5, 9, 2)))

  computeRatio(10, 20, 5) match
    case Right(v)  => println(s"ratio = $v")
    case Left(err) => println(s"error: $err")
